futures-core/src/executor.rs

```diff
@@ -0,0 +1,32 @@
+//! dox
+
+if_std! {
+    use std::boxed::Box;
+    use Future;
+
+    /// TODO: dox
+    pub trait Executor {
+        /// TODO: dox
+        fn spawn(&self, f: Box<Future<Item = (), Error = ()> + Send>) -> Result<(), SpawnError>;
```

Comment:
We had discussed making this `&mut self` on the RFC-- I think this would allow you to get rid of the `RefCell` in `LocalPool`.

Reply:
Then you can't implement `Executor` on an `Arc<E: Executor>`.

Also, we can't get rid of the `RefCell`, because the handles are behind an `Rc`.

Reply:
I just did a big complicated dance to satisfy myself that I couldn't make it work-- the piece I was missing is that you want to be able to `spawn_local` from inside `Future::poll`, while also holding onto a `task::Context`.

Reply:
@aturon fwiw, the doc did switch to `&mut self`.

Is being able to impl `Executor` on `Arc<E: Executor>` a requirement? What is the use case? The executor can always provide a cloneable handle.

Reply:
Either way, `&mut self` vs. `&self` isn't a critical distinction.

Reply:
I don't think the `Executor` for `Arc<E: Executor>` is the critical piece: the important part is that this needs to work somehow:

```rust
let local_pool = LocalPool::new();
let local_exec = local_pool.executor();
local_pool.run_until(|f| future
    .and_then(|_| local_exec.spawn_local(some_other_task))
    .and_then(...));
```

`local_exec` needs to be accessible from inside the `run_until` method of `local_pool`.

Reply:
@cramertj Why does that depend on `&mut self` vs. `&self` on the executor trait?

It [looks like](https://github.com/rust-lang-nursery/futures-rs/pull/763/files#diff-9c6376c5ce46d9e39358812a3269ab6bR60) `executor()` returns a handle by value backed by an `Rc`, so it should be able to be moved into the closure.

You can also impl `Executor` on `&LocalExecutor` to be able to spawn with an immut ref to the local executor.